### ISSUE TITLE
Numpy Adaptor File Structure Issue

### ISSUE THREAD

**Issue by rlizzo** [vmtk/vmtk#193]
Hey @lantiga,

In my use of the new numpy adaptor modules, I've come across a very minor bug that I wanted to get your input about before fixing it. 

In the prior PR for the numpy adaptor feature https://github.com/vmtk/vmtk/pull/181, we made the decision to switch from a dictionary to a list storage method for the `CellPointIds` in vmtkCenterlinesToNumpy. This was the only part of the `ArrayDict` data structure (for images, surfaces, and centerlines) that stores the numpy arrays as elements in a python list rather than key/value pairs in a dictionary.

See:
https://github.com/vmtk/vmtk/blob/edfe00b3e7decc38e914c200201b983861674998/vmtkScripts/contrib/vmtkcenterlinestonumpy.py#L98-L107

This was due to your feedback (which i've coppied/pasted below)

> I would suggest not to use a dictionary after 'CellPointIds', but a simple list, that you can index with an integer. You could use an integer as a key to the dictionary also, but there's no reason to do that, since `cellIds` are sequential and start from `0`. In any case, the `str(cellId)` doesn't really look good and I'd avoid it.

While I still agree with your feedback, this has become a problem now that we have implemented saving in hdf5 files. The issue is that if a file is written out to a hdf5 file, the data is preserved, but the data structure type is lost. If we write out a numpy centerlines file to hdf5, and then read it back in with the vmtkNumpyReader script, the recursive load function does not know to store the `cellPointIds` subgroup into a python list (rather than a dict) in the case where the object is a centerline.

Then, if we convert that back to a vtk centerline object using  

https://github.com/vmtk/vmtk/blob/edfe00b3e7decc38e914c200201b983861674998/vmtkScripts/contrib/vmtknumpytocenterlines.py#L92-L102

**_The algorithm recieves a `dict` rather than a `list` to iterate over as it inserts its cells. This breaks the code._** 

The simplest solution is to just use a dict to store the `cellPointIds` arrays. While i agree that using `str(cellId)` is not the cleanest bit of code, it may be worth it, as this change would allow all subcomponents of the `arrayDict` structure to use dicts instead of having a list used in this one specific instance. 

The other solution (which could complicate the API quite a bit) is to use hdf5 attributes to tag the original dtype of the data structure, and check the attribute to determine the struct to use as we iterate over the hdf5 groups. I'm hesitant to use this because if someone wanted to create a hdf5 file to read into vmtk, they would have to remember to set the attributes correctly (and still would have to remember about this special case for centerlines). 

Hopefully this is clear, i can explain more if needed. just looking for your input before i go and make significant changes to the `arrayDict` structure. Any other solutions you can think of? 

PS: the pickle format does not have this problem as it serializes the entire python object with all type attributes included.

**Comment by lantiga** [comment]
Hi @rlizzo, sorry for the long wait.

I personally would store the cellPointIds in a unique numpy array (not a list of arrays). The array should be structured as:
```
npts_cell0 ptid0 ptid1 .. ptidn npts_cell1 ptid0 ptid1 .. ptidn ...
```

**SIDE NOTE**: use int64 for indices everywhere (I missed this one earlier).

So the code to numpy would become
```
nElements = 0
numberOfCells = clWrapper.VTKObject.GetNumberOfCells() 
for cellId in range(numberOfCells):
    cell = clWrapper.VTKObject.GetCell(cellId) 
    numberOfPointsPerCell = cell.GetNumberOfPoints() 
    nElements += 1 + numberOfPointsPerCell

cellArray = np.zeros(shape=nElements, dtype=np.int64) 

offset = 0
for cellId in range(numberOfCells):
    cell = clWrapper.VTKObject.GetCell(cellId) 
    numberOfCellPoints = cell.GetNumberOfPoints() 
    cellArray[offset] = numberOfPointsPerCell
    for idx in range(numberOfCellPoints): 
        cellArray[offset + 1 + idx] = cell.GetPointId(idx)
    offset += 1 + numberOfCellPoints

self.ArrayDict['CellData']['CellPointIds'] = cellArray
```
the only disadvantage is that we have to iterate once over the cell array just to find out how large the array has to be. The alternative is to get this information directly from the vtk cell array - it has the same structure.

The code for converting backwards becomes
```
for cellKey in self.ArrayDict['CellData'].keys(): 
  
    if cellKey == 'CellPointIds': 
        cellDataArray = vtk.vtkCellArray()
        cellArray = self.ArrayDict['CellData']['CellPointIds']
        offset = 0
        while offset < len(cellArray):
            numberOfCellPoints = cellArray[offset]
            cellDataArray.InsertNextCell(numberOfCellPoints)
            for idx in range(numberOfCellPoints):
                pointId = cellArray[offset + 1 + idx]
                cellDataArray.InsertCellPoint(pointId)
            offset += 1 + numberOfCellPoints
  
        polyData.SetLines(cellDataArray) 
```
I haven't tested this, but it should be close to working. 

Do you think this solves the issue or am I missing something?

**Comment by rlizzo** [comment]
Thanks for the feedback @lantiga, 

While your solution works, I believe that a single list alone would really hinder the usability of the data structure. I can't help but feel requiring users to accurately keep track of the `offset` (aka `npts_cellx`) index to find which `CellPointIds` indexes correspond to a particular `CellId` is sort-of clunky for a user facing data structure. It would be all to easy for someone to improperly implement a loop/index tracker and have the code break in really odd ways. 

I hope you can understand my concern here - for a module whose whole purpose is to provide this data in an easy to use fashion, it would be out of place. I can't help but feel that there's a more obvious implementation. 

I'm going to ruminate on this for a few days and report back after the weekend - i'm swamped and won't be able to work on it till next week at the earliest anyways. Let me know your thoughts.

**Comment by rlizzo** [comment]
This has been resolved in the 1.4 release
